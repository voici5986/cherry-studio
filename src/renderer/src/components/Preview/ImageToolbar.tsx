import { ResetIcon } from '@renderer/components/Icons'
import { classNames } from '@renderer/utils'
import { ChevronDown, ChevronLeft, ChevronRight, ChevronUp, Scan, ZoomIn, ZoomOut } from 'lucide-react'
import { memo, useCallback } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

import ImageToolButton from './ImageToolButton'

interface ImageToolbarProps {
  pan: (dx: number, dy: number, absolute?: boolean) => void
  zoom: (delta: number, absolute?: boolean) => void
  dialog: () => void
  className?: string
}

const ImageToolbar = ({ pan, zoom, dialog, className }: ImageToolbarProps) => {
  const { t } = useTranslation()

  // 定义平移距离
  const panDistance = 20

  // 定义缩放增量
  const zoomDelta = 0.1

  const handleReset = useCallback(() => {
    pan(0, 0, true)
    zoom(1, true)
  }, [pan, zoom])

  return (
    <ToolbarWrapper className={classNames('preview-toolbar', className)} role="toolbar" aria-label={t('preview.label')}>
      {/* Up */}
      <ActionButtonRow>
        <Spacer />
        <ImageToolButton
          tooltip={t('preview.pan_up')}
          icon={<ChevronUp size={'1rem'} />}
          onClick={() => pan(0, -panDistance)}
        />
        <ImageToolButton tooltip={t('preview.dialog')} icon={<Scan size={'1rem'} />} onClick={dialog} />
      </ActionButtonRow>

      {/* Left, Reset, Right */}
      <ActionButtonRow>
        <ImageToolButton
          tooltip={t('preview.pan_left')}
          icon={<ChevronLeft size={'1rem'} />}
          onClick={() => pan(-panDistance, 0)}
        />
        <ImageToolButton tooltip={t('preview.reset')} icon={<ResetIcon size={'1rem'} />} onClick={handleReset} />
        <ImageToolButton
          tooltip={t('preview.pan_right')}
          icon={<ChevronRight size={'1rem'} />}
          onClick={() => pan(panDistance, 0)}
        />
      </ActionButtonRow>

      {/* Down, Zoom */}
      <ActionButtonRow>
        <ImageToolButton
          tooltip={t('preview.zoom_out')}
          icon={<ZoomOut size={'1rem'} />}
          onClick={() => zoom(-zoomDelta)}
        />
        <ImageToolButton
          tooltip={t('preview.pan_down')}
          icon={<ChevronDown size={'1rem'} />}
          onClick={() => pan(0, panDistance)}
        />
        <ImageToolButton
          tooltip={t('preview.zoom_in')}
          icon={<ZoomIn size={'1rem'} />}
          onClick={() => zoom(zoomDelta)}
        />
      </ActionButtonRow>
    </ToolbarWrapper>
  )
}

const ToolbarWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  position: absolute;
  gap: 4px;
  right: 1em;
  bottom: 1em;
  z-index: 5;

  .ant-btn {
    line-height: 0;
  }
`

const ActionButtonRow = styled.div`
  display: flex;
  justify-content: center;
  gap: 4px;
  width: 100%;
`

const Spacer = styled.div`
  flex: 1;
`

export default memo(ImageToolbar)
