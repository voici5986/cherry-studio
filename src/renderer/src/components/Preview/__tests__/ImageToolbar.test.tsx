import { fireEvent, render, screen } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import ImageToolbar from '../ImageToolbar'

// Mock react-i18next
vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => key
  })
}))

// Mock ImageToolButton
vi.mock('../ImageToolButton', () => ({
  default: vi.fn(({ tooltip, onClick, icon }) => (
    <button type="button" onClick={onClick} role="button" aria-label={tooltip}>
      {icon}
    </button>
  ))
}))

// Mock lucide-react icons
vi.mock('lucide-react', () => ({
  ChevronUp: () => <span data-testid="chevron-up">↑</span>,
  ChevronDown: () => <span data-testid="chevron-down">↓</span>,
  ChevronLeft: () => <span data-testid="chevron-left">←</span>,
  ChevronRight: () => <span data-testid="chevron-right">→</span>,
  ZoomIn: () => <span data-testid="zoom-in">+</span>,
  ZoomOut: () => <span data-testid="zoom-out">-</span>,
  Scan: () => <span data-testid="scan">⊞</span>
}))

vi.mock('@renderer/components/Icons', () => ({
  ResetIcon: () => <span data-testid="reset">↻</span>
}))

// Mock utils
vi.mock('@renderer/utils', () => ({
  classNames: (...args: any[]) => args.filter(Boolean).join(' ')
}))

describe('ImageToolbar', () => {
  const mockPan = vi.fn()
  const mockZoom = vi.fn()
  const mockOpenDialog = vi.fn()

  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('should match snapshot', () => {
    const { asFragment } = render(<ImageToolbar pan={mockPan} zoom={mockZoom} dialog={mockOpenDialog} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('calls onPan with correct values when pan buttons are clicked', () => {
    render(<ImageToolbar pan={mockPan} zoom={mockZoom} dialog={mockOpenDialog} />)

    fireEvent.click(screen.getByRole('button', { name: 'preview.pan_up' }))
    expect(mockPan).toHaveBeenCalledWith(0, -20)

    fireEvent.click(screen.getByRole('button', { name: 'preview.pan_down' }))
    expect(mockPan).toHaveBeenCalledWith(0, 20)

    fireEvent.click(screen.getByRole('button', { name: 'preview.pan_left' }))
    expect(mockPan).toHaveBeenCalledWith(-20, 0)

    fireEvent.click(screen.getByRole('button', { name: 'preview.pan_right' }))
    expect(mockPan).toHaveBeenCalledWith(20, 0)
  })

  it('calls onZoom with correct values when zoom buttons are clicked', () => {
    render(<ImageToolbar pan={mockPan} zoom={mockZoom} dialog={mockOpenDialog} />)

    fireEvent.click(screen.getByRole('button', { name: 'preview.zoom_in' }))
    expect(mockZoom).toHaveBeenCalledWith(0.1)

    fireEvent.click(screen.getByRole('button', { name: 'preview.zoom_out' }))
    expect(mockZoom).toHaveBeenCalledWith(-0.1)
  })

  it('calls onReset with correct values when reset button is clicked', () => {
    render(<ImageToolbar pan={mockPan} zoom={mockZoom} dialog={mockOpenDialog} />)

    fireEvent.click(screen.getByRole('button', { name: 'preview.reset' }))
    expect(mockPan).toHaveBeenCalledWith(0, 0, true)
    expect(mockZoom).toHaveBeenCalledWith(1, true)
  })

  it('calls onOpenDialog when dialog button is clicked', () => {
    render(<ImageToolbar pan={mockPan} zoom={mockZoom} dialog={mockOpenDialog} />)

    fireEvent.click(screen.getByRole('button', { name: 'preview.dialog' }))
    expect(mockOpenDialog).toHaveBeenCalled()
  })
})
