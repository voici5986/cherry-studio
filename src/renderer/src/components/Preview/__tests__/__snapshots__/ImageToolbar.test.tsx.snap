// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageToolbar > should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  position: absolute;
  gap: 4px;
  right: 1em;
  bottom: 1em;
  z-index: 5;
}

.c0 .ant-btn {
  line-height: 0;
}

.c1 {
  display: flex;
  justify-content: center;
  gap: 4px;
  width: 100%;
}

.c2 {
  flex: 1;
}

<div
    aria-label="preview.label"
    class="c0 preview-toolbar"
    role="toolbar"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
      <button
        aria-label="preview.pan_up"
        role="button"
        type="button"
      >
        <span
          data-testid="chevron-up"
        >
          ↑
        </span>
      </button>
      <button
        aria-label="preview.dialog"
        role="button"
        type="button"
      >
        <span
          data-testid="scan"
        >
          ⊞
        </span>
      </button>
    </div>
    <div
      class="c1"
    >
      <button
        aria-label="preview.pan_left"
        role="button"
        type="button"
      >
        <span
          data-testid="chevron-left"
        >
          ←
        </span>
      </button>
      <button
        aria-label="preview.reset"
        role="button"
        type="button"
      >
        <span
          data-testid="reset"
        >
          ↻
        </span>
      </button>
      <button
        aria-label="preview.pan_right"
        role="button"
        type="button"
      >
        <span
          data-testid="chevron-right"
        >
          →
        </span>
      </button>
    </div>
    <div
      class="c1"
    >
      <button
        aria-label="preview.zoom_out"
        role="button"
        type="button"
      >
        <span
          data-testid="zoom-out"
        >
          -
        </span>
      </button>
      <button
        aria-label="preview.pan_down"
        role="button"
        type="button"
      >
        <span
          data-testid="chevron-down"
        >
          ↓
        </span>
      </button>
      <button
        aria-label="preview.zoom_in"
        role="button"
        type="button"
      >
        <span
          data-testid="zoom-in"
        >
          +
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
